### ISSUE TITLE
Add method to create an initial config file

### ISSUE THREAD

**Issue by twpayne** [twpayne/chezmoi#241]
In #170, @fgrosse wrote:

> Maybe in the future it could also be extended to ask the user for parameters to automatically generate ~/.config/chezmoi/chezmoi.yaml based on some definition in the source repo but I would not go that far yet and instead defer this topic, maybe into a later issue.

This is that later issue for discussion.

**Comment by twpayne** [comment]
Here's one way of doing it. @fgrosse, I'd be interested in your thoughts.

The template code below is untested, the actual syntax might be slightly different.

In the source directory, if the user has a file called `.chezmoi.yaml.tmpl` then, if no config file already exists, then `chezmoi init` will execute this file as a template and write the output to a new config file in the default location (`.config/chezmoi/chezmoi.yaml`).

We can use a template function to prompt the user for values, for example:

```
data:
  email: "{{ promptString "email" }}"
```

where we define `promptString` as a template function that prompts the user for a string value.

If the same value is needed in multiple places, then we can use a variable:

```
{{ $email := promptString "email" }}
data:
  email: "{{ $email }}"
  mailtoURL: "mailto:{{ $email }}"
```

We'll also need to check for other supported config file formats, e.g. `.chezmoi.toml.tmpl`, `.chezmoi.json.tmpl`, etc., as well as copying the config file verbatim if it doesn't have a `.tmpl` extension.

This approach should provide all the desired functionality while sharing a lot of common with normal `chezmoi` usage.

Thoughts?

**Comment by fgrosse** [comment]
I was just about to open the issue :sweat_smile: 
The idea of using a template to generate this file is quite nice (and bit templateception :sunglasses:).
This provides very good flexibility and is nicely consistent with the way things work in chezmoi.

Some general thoughts:
- Just for clarification so we are on the same page: rendering the config template would be executed by `chezmoi init`, after the checkout but before creating the dotfiles. 
- Should there be a flag to skip this step? Probably no because if you are using templates with custom variables the apply step of the init command will fail with an error anyway (i.e. that's the problem that motivated this request).
- Can it be useful to define variables via environment variables so this remains scriptable and no interactive prompts are strictly required in the init command?
- If later required we can easily add new functions that can be used in the template. However this should be done with care to not bloat this part to much. Anyway just a thought as I'm pretty sure you'll have an eye on this.

I'm gonna think about it a bit more and try to come up with different use cases that I have in mind but so far I agree that this should provide all functionality I can see so far plus its probably easy to extend if we need to. :+1:

**Comment by twpayne** [comment]
> Just for clarification so we are on the same page: rendering the config template would be executed by chezmoi init, after the checkout but before creating the dotfiles.

Yes, exactly.

> Should there be a flag to skip this step? Probably no because if you are using templates with custom variables the apply step of the init command will fail with an error anyway (i.e. that's the problem that motivated this request).

Agree, no flag initially.

> Can it be useful to define variables via environment variables so this remains scriptable and no interactive prompts are strictly required in the init command?

Interesting idea, I'm not sure, probably not. I would expect most people will not use many configuration variables (I use just three myself). `chezmoi` is for managing your pets, not your cattle, so if you are scripting `chezmoi`, you should probably be using a configuration management system instead.

Thanks for the feedback!

**Comment by twpayne** [comment]
Re environment variables, let's definitely keep these out in the short term.

In the medium term, templates in `chezmoi` have the hermetic functions from [sprig](http://masterminds.github.io/sprig/) thanks to @jdotdot in #184 . For the generation of the config file, we could consider allowing the non-hermetic functions like [`env`](http://masterminds.github.io/sprig/os.html) as well. This allows a large amount of the environment variable/scripting logic to be implemented in templates, rather than in `chezmoi` itself. This makes sense to me as generating the config file happens exactly once (when running `chezmoi init`) so using non-hermetic functions is less bad.

Given that you've contributed code to `chezmoi` in the past (notably #174) I'll assume that you'll work on this. If you don't have time then please tell me here.

**Comment by fgrosse** [comment]
Sounds all good to me and I would be happy to work on this issue. I won't have time for it though this weekend but hopefully I can make some time next week or the weekend after. The issue is interesting but since its a new feature I don't think it's highly urgent. If you decide you do not want to wait for me and work on it yourself that's of course also fine with me.

How about I write again shortly before I actually start any implementation so you know that as long as I didn't write this is still free for everybody :)

**Comment by fgrosse** [comment]
Quick update: I found some time actually today and started to work on the issue :sweat_smile:

**Comment by twpayne** [comment]
Fixed by @fgrosse in #243.
